### DISCUSSION TITLE
I'm going to pay a programmer to add additional features - would like to coordinate with you

### DISCUSSION THREAD

**Discussion by ShiroiKuma0** [arianneorpilla/jidoujisho#276] [General]
I'm going to pay a programmer to program additional features - that I'd like to see in jidoujisho - and would like to coordinate this with.

I'd like to not fork the project - so designed the features so that they only complement, not undo or something, what you have currently. That way you could still be in charge and we'd keep the whole code here in your source tree.

Questions:

- Would this be OK for you?
- Do you have a friend or recommendation who - which programming services - site to use to hire, so you could work with them - have them submit code modifications, review it if necessary, quickly publish?
- Maybe some developer reading here could volunteer to do this - as a paid job?

If need be - we can discuss directly via email, but maybe beneficial to post here.

The key for me is quick - generally OK code, so no extreme garbage with ugly hacks.

If not feasible I'll fork it - and have someone code it and publish elsewhere.

The current list of changes I've identified:

Media tab - the player window - while playing a video:
	Bottom bar, rightmost icon:
		- add Controls section (inspired by the "Controls" section from Ankidroid's Settings)
this section will list control actions to be performed - with their respective gestures and hotkeys (to facilitate using phone hardware keys and/or an attached external keyboard)
			actions list:
				- toggle play/pause
				- replay current subtitle
				- back (will first skip to beginning of current subtitle, then on second action, to previous subtitle and so on)
				- forward (will go to next subtitle)
				- back and forward audio - by a fixed number of seconds
				- seconds setting for the above back/forward
				- vol up
				- vol down
				- brightness up
				- brightness down
				- toggle display of subtitles
				- share current subtitle (will run the Android share dialog - so you can output the whole sentence to Google Translate for instance)
				- ChatGPT question (default: "Traslated to English: ")
				- share the current subtitle to ChatGPT - switch to Reader/ChatGPT source and paste the ChatGPT question followed by the current subtitle, and send - so that ChatGPT answers.
				- copy current subtitle to clipboard
			each above action will have a default gesture/s, for instance:
				- toggle play/pause: tap screen center
			each above action's gestures can be edited - deleted, new gestures/hotkeys added - with a warning if the gesture/hotkey is already added to another action (specifying what the action is) - each action can have unlimited gestures/hotkeys
			possible gestures:
				- key (will be keyboard keys, and hardware keys - VolUp etc.)
				- gestures:
					touch, long touch, double touch, swipe (1, 2, 3, 4, and 5 fingers)
				- areas for gestures (where will be touched or where swipe starts)
					top left, top center, top right, center left, center, center right, bottom left, bottom center, bottom right
				- direction for swipe:
					left, right, up, down
	Bottom bar:
		- add toggle permanent visibility on/off button
		- subtitle back and forward buttons
		- audio back and forward buttons
		- vol up and down buttons
		- brightness up and down buttons
		Bottom bar, second icon from the right: Subtitle Timing and Appearance:
			- add external font TTF/OTF selection from disc
			- add font color selection - RGB dialog
			- add font weight selection
			- add subtitle outline color - RBG dialog
Reader tab:
	Ebook source:
		- Display bottom bar - similar to video player - when touched on the bottom - here add same Controls and Font Settings possibilities
		- Add a possibility to add an audiobook (of the book being read) - so you can load an mp3 file here and have it play/pause rewind while reading 
		- Add some buttons on this bar - on the left side - to achieve the above:
toggle fixed display of this bar - so it doesn't disappear - when you want to do a lot of playing-rewinding.
			- play/pause audio
			- vol up
			- vol down
			- back in audio
			- forward in audio
			- open audio (to select the file, save the file then to some setting, so that it's opened automatically with the book)
			- send current sentence to ChatGPT - the position of the current sentence will be determined by last click - so needs to be saved somewhere when clicking the text to display translations.
			- share current sentence
			- copy current sentence
	ChatGPT source:
		- add a button to the right of the send button on the bottom right - to go back to the video, if sent here from the playing video - to go back to the read book, if sent here from the Reader epub

**Comment by arianneorpilla** [comment]
I think what you want to do is pay a bounty for certain features be implemented. [**There are certain services for that kind of arrangement in open source**](https://bountysource.com/). For myself, I personally accept donations via **[Ko-Fi](https://ko-fi.com/lrorpilla)**, but I generally like to keep this as a hobby project and I don't take payment in exchange for being able to guarantee that certain changes will be made. If it's another developer, as long as it's explicit and you are able to communicate that you are the individual offering the bounty and that me as the project maintainer does not owe another developer payment, it should be fine.

Additionally, [**this project is licensed under GPLv3**](https://github.com/lrorpilla/jidoujisho/blob/main/LICENSE), so as long as you're conforming with the terms of the license that is alright. I'm happy to take and review pull requests for code changes.

I just want to ensure that all code changes made to this project will always be free and open with source available for the community.

Thanks again and I really appreciate your interest, @ShiroiKuma0. I'm a little busy with life at the moment but you can always just get in touch if you need to clarify anything else or if you want to collaborate.

**Reply by ShiroiKuma0** [reply]
The GPLv3 is understood, I wouldn't have it any other way.

I wasn't thinking of a bounty - as the way I understand bounties like this work - for instance on the site you mention :is you say “I'll pay $200 for this and give a specification.” This then might take long or forever, since noone is interested. Or you split it into tasks - and it seems it might take forever. I'd really like to have the full set done.

Or am I wrong.

I was thinking more of: these are tha changes I want, then developers say how much they want for it, and cca how long it'll take - then we agree and they do it.

I know in corp world there's markets like this - usually Indian developers, not epensive. The problem there is it's not GPL, so they might not be used to complying with everything. .That's why their submission would have to be checked that it's not super messy, builds ebc.

So I thought I might try to get a developer on this forum to bid, I'd accept the price and then we might split it in stages - for each part that I list:

- I don't know what he'd have to sign - submit to you, so we protect the code so that GP.L isn't contaminated - and all of a sudden, later, somebody appears saying this software is not Free - do you have an idea?
- he'd code it
- you'd pull it 
- you'd check it does what he says
- you'd build it and publish
- I'd pay him for the step
- and we'd go on.

Sound feasible? Or a bounty is really more doable somehow?

**Reply by arianneorpilla** [reply]
I personally won't really get involved with how the pull requests come to the project. That will be up to you and I am just going to be receiving code changes to approve. If needed, I will with reasonable effort, give advice on how to make certain changes if necessary.

Whether you decide to use an open source bounty service, a freelancer service like Fiverr or privately handpick and employ an individual is up to you, and everything else other than the changes being accepted by the project is really a private matter between you and whoever you decide to pick. You can also probably wait to see if someone is interested to implement these features in exchange for payment from seeing this.

I would really do some studying up on what this entire endeavour would entail before you proceed with anything.

**Comment by ShiroiKuma0** [comment]
Right — I didn't realize, the copyright is simple, if the programmer will submit to you to upstream via github, in line with conditions, that should take care of all the legal stuff etc. 

I'll make an effort to find somebody apt now. 

I was thinking of freelancer.com or guru.com — you've mentioned two sites. Do you have any other pointers where I should look or think about?

**Comment by ShiroiKuma0** [comment]
Copying email communication thread:
---------------

This is ShiroiKuma0 from github — I have found a programmer on Freelancer
who will program the updates envisioned by me and push them to your
repository on github. I've split the desired changes into 6 blocks which
will be done each as a piece — all of them within 10 days. I should
conclude the contract with him on Freelancer today.

Do you think you'll be able to check after every block he pushes
upstream, and potentially build a new version?

Do you want to coordinate anything beforehand or talk online or
something? :@)

The scope of the updates — as we agreed them on Freelancer:

1.) Add the posibility to select subtitle font, its color, and outline
color - when playing videos. (Currently, fonts can be "selected" based on
their Google name or something - don't use that. I mean select a local
otf/ttf file on the phone. Currently only the font size can be selected
in below settings, not color and outline color.)

Media tab - the player window - while playing a video:
Bottom bar, second icon from the right: Subtitle Timing and Appearance:
    - add external font TTF/OTF selection from disc
    - add font color selection - RGB dialog
    - add font weight selection
    - add subtitle outline color - RBG dialog

2.) Modify the bottom bar that currently shows when playing a video and
the screen is touched, to enable easier video operation.

Media tab - the player window - while playing a video:
Bottom bar - on its left side:
    - add toggle permanent visibility on/off button
    - subtitle back and forward buttons
    - audio back and forward buttons (to skip by a certain number of
seconds, default 5, this number to be then modified in Settings, as per
the further module below)
    - vol up and down buttons
    - brightness up and down buttons

3.) Add a bottom bar also for when reading an epub and add the same font
color etc. settings to select when reading an epub.

Reader tab:
Ebook source:
    - Display bottom bar - similar to video player - when touched on the
bottom - here add same Controls and Font Settings possibilities

4.) Add a possibility to add an audiobook (of the book being read) in the
epub reader - so you can load an mp3 file here and have it play/pause
rewind while reading the book. Implement this in the bottom bar from step
before.

Add some buttons on the bottom bar - on the left side - to achieve the
above:
    - toggle fixed display of this bar - so it doesn't disappear - when
you want to do a lot of playing-rewinding.
    - play/pause audio
    - vol up
    - vol down
    - back in audio
    - forward in audio
    - open audio (to select the file, save the file then to some setting,
so that it's opened automatically with the book)
    - send current sentence to ChatGPT - the position of the current
sentence will be determined by last click - so needs to be saved
somewhere when clicking the text to display translations.
    - share current sentence
    - copy current sentence

5.) Add a button in the ChatGPT screen to quickly take you back to the
video or the book you came from.

ChatGPT source:
    - add a button to the right of the send button on the bottom right -
to go back to the video, if sent here from the playing video - to go back
to the read book, if sent here from the Reader epub

6.) Add programmable controls with gestures and hotkeys for Media tab
when playing videos. To be done last as this is bigger - and more
significant - inspired by Ankidroid - install it and check it out, you'll
immediately see what I mean - when you look at the "Controls" section in
Ankidroid's Settings.

Media tab - the player window - while playing a video:
Bottom bar, rightmost icon:
    - add Controls section (inspired by the "Controls" section from
Ankidroid's Settings)
this section will list control actions to be performed - with their
respective gestures and hotkeys (to facilitate using phone hardware keys
and/or an attached external keyboard)

actions list:
    - toggle play/pause
    - replay current subtitle
    - back (will first skip to beginning of current subtitle, then on
second action, to previous subtitle and so on)
    - forward (will go to next subtitle)
    - back and forward audio - by a fixed number of seconds
    - seconds setting for the above back/forward
    - vol up
    - vol down
    - brightness up
    - brightness down
    - toggle display of subtitles
    - share current subtitle (will run the Android share dialog - so you
can output the whole sentence to Google Translate for instance)
    - ChatGPT question (default: "Traslated to English: ")
    - share the current subtitle to ChatGPT - switch to Reader/ChatGPT
source and paste the ChatGPT question followed by the current subtitle,
and send - so that ChatGPT answers.
    - copy current subtitle to clipboard

each above action will have a default gesture/s, for instance:
    - toggle play/pause: tap screen center

each above action's gestures can be edited - deleted, new
gestures/hotkeys added - with a warning if the gesture/hotkey is already
added to another action (specifying what the action is) - each action can
have unlimited gestures/hotkeys
possible gestures:
    - key (will be keyboard keys, and hardware keys - VolUp etc.)
    - gestures:
        touch, long touch, double touch, swipe (1, 2, 3, 4, and 5
fingers)
    - areas for gestures (where will be touched or where swipe starts)
        top left, top center, top right, center left, center, center
right, bottom left, bottom center, bottom right
    - direction for swipe:
        left, right, up, down

--------------
lrorpilla response:
--------------

Hi Shiroikuma,

The only thing I find objectionable about the feature list is the deeper ChatGPT integration. I really do not recommend doing that and I don't think it is a good look for the project to promote such a feature for use as a reliable material for translation and reference.

If you decide to move forward with such a feature, I'm going to have to ask you to include a disclaimer with new ChatGPT features that relying on it is not optimal and may actively cause harm to a user's learning. Of course at the end of the day, what you request with your code changes is up to you.

The reason why I haven't included such features is because I actually want to discourage users from relying on them. I'm of the opinion users should outgrow/avoid ChatGPT and never need them, rather than being spoiled with options.

Other than that, I will try to get to your pull requests as I receive them. I will ensure that the features in the pull requests organically fit into the app as before approving them. Of course, if a feature does not align with what I deem to be in the best interests of the project, it can always exist in a separate fork -- but I'll be reasonable in that regard.

Again, I really appreciate the interest and your resolve in going forward with this endeavour, it means a lot to me that people care enough to contribute to this project.

If you will, would you include this e-mail in the discussions post that you've already made on GitHub? I prefer to keep these communications transparent for the entire community to see.

**Comment by ShiroiKuma0** [comment]
I don't see his updates published - he's implemented font selection, font color etc. in the media player and other stuff.

He was supposed to send pull request to your repo - are you reviewing the changes or it didn't happen?

**Reply by arianneorpilla** [reply]
I work full-time these days. I work, take care of myself, have time for my hobbies. I will review and test the changes when I feel like it, as these changes aren't critical changes and are extra features, they are very low priority for me, unlike the other pull request that was merged today where existing users were having problems.

I would say just maintain a fork in the separate repo for now and let the pull requests add up per set of features as necessary. I can't approve these as they arrive and I am very particular with UX when it comes to new features so I will have to ensure that whatever I accept into the project fits with everything else.

I would just maintain the fork for now and continue with development so that you aren't blocked when I'm busy with life.

I will get to them eventually.

**Comment by ShiroiKuma0** [comment]
I wanted him to also implement font color change in epub reader - but you're using the ettu reader and he says:

But as current project is product mode, can't modify anything.
Only one way to modify is to modify original project and re-build the project.
But now this project doesn't work.

Would it be possible just to hardcode the font color for night mode for text as yellow (not white) for me to build on the side? As per the above communication he says he can't do it?

**Reply by arianneorpilla** [reply]
The source code for ttu can be found at https://github.com/ttu-ttu/ebook-reader

**Comment by Lyhcan** [comment]
I would love to see those features implemented someday

**Comment by xrishox** [comment]
having more robust gesture support definitely would be cool. jumping back and fourth between subtitles with a gesture would be killer.

**Comment by ShiroiKuma0** [comment]
So I had the following implemented:
    Move back and forth between subtitles while playing video.
    Select subtitle font, color, select outline color.
    Hardware keys playback control - b p f - back, play/pause, forward.
    Brightness volume control.
    Bottom bar toggleable always on/off.
    Some other settings.

It's synched to 2.8.8 jidoujisho version - unfortunatelly, I can't build it anymore and cannot dedicate the time now to clean and push here to the git - an not git proficient enough to do it in a reasonable amount of time. Anyhow the source is at: https://github.com/ShiroiKuma0/jidoujisho2

And I was able to build an arm64 version - which is at, if anyone wants to use it: https://github.com/ShiroiKuma0/jidoujisho2/releases/tag/v2.8.8
https://github.com/ShiroiKuma0/jidoujisho2/releases/download/v2.8.8/app.lrorpilla.yuuna.apk

I'm using it now - works well - the jump back and forth between subtitles is a killer, also the hardware keys programmed it - I have tasker actions remapped to volume keys while in jidoujisho to send these keys, don't have to fiddle with the onscreen controls.

This is enough for me, can't integrate sensibly the code here - it's very useful though. So if someone can merge it here so it floats into the code here, might be beneficial.
